haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs

```diff
@@ -246,96 +245,61 @@ declsFromRenamed ctx (hsGroup, _, _, _) =
-                                typeStringyType
-            Implicit  -> implicitCase
-            Qualified -> implicitCase
-                -- Quoting Note [HsForAllTy tyvar binders]:
-                --
-                -- "Qualified currently behaves exactly as Implicit, but it is
-                -- deprecated to use it for implicit quantification."
-      where
-        implicitCase =
-            let boundTyVarNames = Set.fromList
-                                . map (hsTyVarBinderName . unLoc) . hsq_tvs
-                                $ binders
-                usages = mapMaybe (interestingNameToLoc boundTyVarNames)
-                       . universeBi
-                       $ (ty, context)
-                firstUsages = mapMaybe keepFirst
-                            . groupBy ((==) `on` fst)
-                            . sortBy (comparing fst)
-                            $ usages
-            in map mkDecl firstUsages
-                 where mkDecl (n,l) = declWithWrappedIdLoc typeStringyType
-                                                           (L l n)
-        --
-        interestingNameToLoc :: Set.Set Name -> LHsType Name
-                             -> Maybe (Name, SrcSpan)
-        interestingNameToLoc s (L loc (hsTypeVarName -> Just n))
-            | Set.member n s = Just $! (n, loc)
-            | otherwise = Nothing
-        interestingNameToLoc _ _ = Nothing
-    declsFromForall _ = []
-#endif
     -- | For typevar binders directly attached to datatype declarations (like
     -- data, type, class etc), the location of the name is correctly set to
     -- the span of the type variable.
     -- TODO(robinpalotai): as in declsFromForall about stripping kind signature
     --   from span.
-#if __GLASGOW_HASKELL__ >= 800
     hsqShim = hsq_explicit
     declsFromDataBinders :: LHsQTyVars GhcRn -> [DeclAndAlt]
-#else
-    hsqShim = hsq_tvs
-    declsFromDataBinders :: LHsTyVarBndrs Name -> [DeclAndAlt]
-#endif
     declsFromDataBinders = map (mkDecl . hsTyVarBinderName . unLoc) . hsqShim
       where
         mkDecl n = nameDeclAlt ctx n Nothing typeStringyType
     hsTyVarBinderName :: HsTyVarBndr id -> IdP id
-    hsTyVarBinderName = \case
-        HsTyVarBndrCompat n -> mayUnLoc n
+    hsTyVarBinderName (UserTyVar _ n) = unLoc n
+    hsTyVarBinderName (KindedTyVar _ n _) = unLoc n
+    hsTyVarBinderName (XTyVarBndr _) =
+        error "should not hit XTyVarBndr when accessing renamed AST"
     -- Datatypes.
     dataDecls :: LTyClDecl GhcRn -> [DeclAndAlt]
-    dataDecls (L _ (DataDeclCompat locName binders defn)) =
+    dataDecls (L _ (DataDecl _ locName binders _ defn)) =
         let top = dataCtorLikeDecl locName
             ctors = map (conDecls . unLoc) . dd_cons $ defn
             tyvars = declsFromDataBinders binders
         in top:(ctors ++ tyvars)
       where
         conDecls cd =
-            let conLName = head (conDeclNames cd)  -- TODO(robinp): all ctors
```

Comment:
Keep comment?  It seems like it might still be applicable for GADTs.  (Maybe we should file an issue, if there isn't already one?)

Reply:
Done. Also filed an issue for that.